lib/AST/LayoutConstraint.cpp

```diff
@@ -114,4 +136,61 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
 
 SourceRange LayoutConstraintLoc::getSourceRange() const { return getLoc(); }
 
+LayoutConstraint LayoutConstraint::merge(LayoutConstraint Other) {
+  if (!(*this)->isKnownLayout())
+    return Other;
+
+  if ((*this)->isAddressOnlyTrivial() && Other->isKnownSizeTrivial()) {
+    // Prefer a more concrete _Trivial constraint.
+    return Other;
+  }
+
+  if (Other->isAddressOnlyTrivial() && (*this)->isKnownSizeTrivial()) {
+    // Prefer a more concrete _Trivial constraint.
+    return *this;
+  }
+
+
+  if ((*this)->isRefCountedObject() && Other->isNativeRefCounted()) {
+    // Prefer a more concrete _NativeRefCounted constraint.
+    return Other;
+  }
+
+  if (Other->isRefCountedObject() && (*this)->isNativeRefCounted()) {
+    // Prefer a more concrete _NativeRefCounted constraint.
+    return *this;
+  }
+
+  if ((*this)->isClass() && Other->isNativeClass()) {
+    // Prefer a more concrete _NativeClass conttraint.
+  }
+
+  if (Other->isClass() && (*this)->isNativeClass()) {
+    // Prefer a more concrete _NativeClass conttraint.
+  }
+
+  return LayoutConstraint::getUnknownLayout();
```

Comment:
@DougGregor Please have a look at the update that I pushed. It uses a 2-D table, but I'm not quite happy with it. The initialization of the table looked pretty ugly. I had to use macros to shorten it. And I'm still not quite convinced it is really easier. WDYT?